File: Stage19/C/Problem4779.c
#include <stdio.h>

char arr[600000];

void cantor(int start, int size) {
    if (size == 1) return;

    int len = size / 3;

    for (int i = start + len; i < start + len * 2; i++) {
        arr[i] = ' ';
    }

    cantor(start, len);
    cantor(start + len * 2, len);
}

int main(void) {
    int N;

    while (scanf("%d", &N) != EOF) {
        int size = 1;
        for (int i = 0; i < N; i++)
            size *= 3;

        for (int i = 0; i < size; i++)
            arr[i] = '-';

        cantor(0, size);

        for (int i = 0; i < size; i++)
            putchar(arr[i]);
        putchar('\n');
    }

    return 0;
}

File: Stage19/Java/Problem4779.java
package Stage19.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Problem4779 {
    static char[] arr;

    public static void cantor(int start, int size) {
        if (size == 1) return;

        int len = size / 3;

        for (int i = start + len; i < start + len * 2; i++) {
            arr[i] = ' ';
        }

        cantor(start, len);
        cantor(start + len * 2, len);
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        String input;

        while ((input = br.readLine()) != null) {
            int N = Integer.parseInt(input);
            int size = (int) Math.pow(3, N);
            arr = new char[size];

            for (int i = 0; i < size; i++) {
                arr[i] = '-';
            }

            cantor(0, size);

            sb.append(arr).append("\n");
        }

        System.out.println(sb);
        br.close();
    }
}
